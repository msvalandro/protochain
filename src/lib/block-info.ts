export interface BlockInfo {
  index: number
  previousHash: string
  data: string
  difficulty: number
  maxDifficulty: number
  feePerTx: number
}
