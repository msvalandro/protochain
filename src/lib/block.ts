import sha256 from 'crypto-js/sha256'

import { ValidationError } from './validation-error'

interface CreateBlockParams {
  index: number
  previousHash: string
  data: string
}

export class Block {
  private index: number
  private hash: string
  private previousHash: string
  private data: string
  private timestamp: number
  private nonce = 0
  private miner = ''

  constructor({ index, previousHash, data }: CreateBlockParams) {
    this.index = index
    this.previousHash = previousHash
    this.data = data

    this.timestamp = Date.now()
    this.hash = this.generateHash()
  }

  static genesis(): Block {
    const block = new Block({
      index: 0,
      previousHash:
        '0000000000000000000000000000000000000000000000000000000000000000',
      data: 'Genesis Block',
    })

    block.mine(0, 'genesis')

    return block
  }

  private generateHash(): string {
    return sha256(
      this.index +
        this.data +
        this.timestamp +
        this.previousHash +
        this.nonce +
        this.miner,
    ).toString()
  }

  getIndex(): number {
    return this.index
  }

  getHash(): string {
    return this.hash
  }

  private validateData(): void {
    if (!this.data) {
      throw new ValidationError('Invalid block data')
    }
  }

  private validateTimestamp(): void {
    if (this.timestamp < 1) {
      throw new ValidationError('Invalid block timestamp')
    }
  }

  private validatePreviousBlock(hash: string, index: number): void {
    if (hash !== this.previousHash) {
      throw new ValidationError('Invalid previous block hash')
    }

    if (index !== this.index - 1) {
      throw new ValidationError('Invalid block index')
    }
  }

  private validateHash(difficulty: number): void {
    const prefix = '0'.repeat(difficulty)

    if (this.hash !== this.generateHash() || !this.hash.startsWith(prefix)) {
      throw new ValidationError('Invalid block hash')
    }
  }

  private validateMining(): void {
    if (!this.nonce || !this.miner) {
      throw new ValidationError('Block not mined')
    }
  }

  validate(
    previousHash: string,
    previousIndex: number,
    difficulty: number,
  ): void {
    try {
      this.validateData()
      this.validateTimestamp()
      this.validatePreviousBlock(previousHash, previousIndex)
      this.validateMining()

      this.validateHash(difficulty)
    } catch (error) {
      console.error((error as ValidationError).message)
      throw error
    }
  }

  mine(difficulty: number, miner: string): void {
    this.miner = miner
    const prefix = '0'.repeat(difficulty)

    do {
      this.nonce++
      this.hash = this.generateHash()
    } while (!this.hash.startsWith(prefix))
  }
}
