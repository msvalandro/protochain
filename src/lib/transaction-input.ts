import sha256 from 'crypto-js/sha256'
import { ECPairFactory } from 'ecpair'
import * as ecc from 'tiny-secp256k1'

import { ValidationError } from './validation-error'

const ECPair = ECPairFactory(ecc)

interface CreateTransactionInputParams {
  fromAddress: string
  amount: number
  signature: string
}

export class TransactionInput {
  fromAddress: string
  amount: number
  signature: string

  constructor({
    fromAddress,
    amount,
    signature,
  }: CreateTransactionInputParams) {
    this.fromAddress = fromAddress
    this.amount = amount
    this.signature = signature
  }

  generateHash(): string {
    return sha256(this.fromAddress + this.amount).toString()
  }

  sign(privateKey: string): void {
    const keyPair = ECPair.fromPrivateKey(Buffer.from(privateKey, 'hex'))
    const hash = this.generateHash()

    this.signature = keyPair.sign(Buffer.from(hash, 'hex')).toString('hex')
  }

  private validateSignature(): void {
    if (!this.signature) {
      throw new ValidationError('Signature is required')
    }
  }

  private validateAmount(): void {
    if (this.amount < 1) {
      throw new ValidationError('Amount must be greater than 0')
    }
  }

  private validateHash(): void {
    const hash = Buffer.from(this.generateHash(), 'hex')
    const isValid = ECPair.fromPublicKey(
      Buffer.from(this.fromAddress, 'hex'),
    ).verify(hash, Buffer.from(this.signature, 'hex'))

    if (!isValid) {
      throw new ValidationError('Invalid hash')
    }
  }

  validate(): void {
    try {
      this.validateSignature()
      this.validateAmount()
      this.validateHash()
    } catch (error) {
      console.error(error)
      throw new ValidationError(`Invalid transaction input. ${error}`)
    }
  }
}
