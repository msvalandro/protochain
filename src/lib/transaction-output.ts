import sha256 from 'crypto-js/sha256'

import { ValidationError } from './validation-error'

export interface CreateTransactionOutputParams {
  toAddress: string
  amount: number
  txHash?: string
}

export class TransactionOutput {
  private toAddress: string
  private amount: number
  private txHash: string

  constructor({ toAddress, amount, txHash }: CreateTransactionOutputParams) {
    this.toAddress = toAddress
    this.amount = amount
    this.txHash = txHash || ''
  }

  getToAddress(): string {
    return this.toAddress
  }

  getAmount(): number {
    return this.amount
  }

  setTxHash(txHash: string): void {
    this.txHash = txHash
  }

  generateHash(): string {
    return sha256(this.toAddress + this.amount + this.txHash).toString()
  }

  validate(): void {
    if (this.amount < 1) {
      throw new ValidationError('Invalid transaction output amount')
    }
  }
}
