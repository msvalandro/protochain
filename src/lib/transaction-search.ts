import { Transaction } from './transaction'

export interface TransactionSearch {
  transaction: Transaction | undefined
  mempoolIndex: number
  blockIndex: number
}
