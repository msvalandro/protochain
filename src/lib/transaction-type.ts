export enum TransactionType {
  REGULAR = 'regular',
  FEE = 'fee',
}
