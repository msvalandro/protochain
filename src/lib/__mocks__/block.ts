import { TransactionType } from '../transaction-type'
import { Blockchain } from './blockchain'
import { CreateTransactionParams, Transaction } from './transaction'
import { TransactionInput } from './transaction-input'
import { TransactionOutput } from './transaction-output'

interface CreateBlockParams {
  index: number
  previousHash: string
  transactions: Transaction[] | CreateTransactionParams[]
  nonce?: number
  miner?: string
  timestamp?: number
  hash?: string
}

export class Block {
  private index: number
  private hash: string
  private previousHash: string
  private transactions: Transaction[]
  private timestamp: number
  private nonce: number
  private miner: string

  constructor({
    index,
    previousHash,
    transactions,
    nonce,
    miner,
    timestamp,
    hash,
  }: CreateBlockParams) {
    this.index = index
    this.previousHash = previousHash
    this.transactions = transactions.map(
      (tx) => new Transaction({ ...tx } as CreateTransactionParams),
    )
    this.nonce = nonce ?? 0
    this.miner = miner ?? ''

    this.timestamp = timestamp ?? Date.now()
    this.hash = hash ?? this.generateHash()
  }

  static genesis(miner: string): Block {
    const amount = Blockchain.getRewardAmount()

    const transaction = new Transaction({
      type: TransactionType.FEE,
      txOutputs: [new TransactionOutput({ toAddress: miner, amount })],
    })
    transaction.setTransactionOutputHash()

    const block = new Block({
      index: 0,
      previousHash:
        '0000000000000000000000000000000000000000000000000000000000000000',
      transactions: [transaction],
    })

    block.mine()

    return block
  }

  generateHash(): string {
    return 'mock-block-hash'
  }

  getIndex(): number {
    return this.index
  }

  getPreviousHash(): string {
    return this.previousHash
  }

  getTransactions(): Transaction[] {
    // mock invalid tx in block
    if (this.timestamp === 99) {
      return [new Transaction({}), new Transaction({})]
    }

    return [
      new Transaction({
        txInputs: [
          new TransactionInput({
            fromAddress: 'mock-address',
            amount: 100,
          }),
        ],
        txOutputs: [
          new TransactionOutput({ toAddress: 'mock-address', amount: 80 }),
          new TransactionOutput({ toAddress: 'mock-address', amount: 100 }),
        ],
      }),
    ]
  }

  getNonce(): number {
    return this.nonce
  }

  getMiner(): string {
    return this.miner
  }

  getTimestamp(): number {
    return this.timestamp
  }

  getHash(): string {
    return this.hash
  }

  getTransaction(hash: string): Transaction | undefined {
    return this.transactions.find((tx) => tx.getHash() === hash)
  }

  hasTransaction(hash: string): boolean {
    return hash === 'invalid-transaction'
  }

  addTransaction(transaction: Transaction): void {
    this.transactions.push(transaction)
  }

  validate(hash: string): void {
    if (hash === 'invalid-hash') {
      throw new Error('Invalid block')
    }
  }

  mine(): void {}
}
