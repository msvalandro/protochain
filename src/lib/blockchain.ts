import { Block } from './block'
import { ValidationError } from './validation-error'

export class Blockchain {
  private blocks: Block[]
  private nextIndex = 1
  static readonly DIFFICULTY_FACTOR = 5

  constructor() {
    this.blocks = [Block.genesis()]
  }

  getBlock(hash: string): Block | undefined {
    return this.blocks.find((block) => block.getHash() === hash)
  }

  getLastBlock(): Block {
    return this.blocks[this.blocks.length - 1]
  }

  getBlocks(): Block[] {
    return this.blocks.slice()
  }

  getDifficulty(): number {
    return Math.ceil(this.blocks.length / Blockchain.DIFFICULTY_FACTOR)
  }

  addBlock(block: Block): void {
    try {
      const previousBlock = this.getLastBlock()

      block.validate(
        previousBlock.getHash(),
        previousBlock.getIndex(),
        this.getDifficulty(),
      )

      this.blocks.push(block)
      this.nextIndex++
    } catch (error) {
      throw new ValidationError(
        `Invalid block #${block.getHash()}. Error: ${error}`,
      )
    }
  }

  isValid(): boolean {
    for (let i = this.blocks.length - 1; i > 0; i--) {
      const currentBlock = this.blocks[i]
      const previousBlock = this.blocks[i - 1]

      try {
        currentBlock.validate(
          previousBlock.getHash(),
          previousBlock.getIndex(),
          this.getDifficulty(),
        )
      } catch (error) {
        console.error(
          `Invalid block #${currentBlock.getHash()}. Error: ${error}`,
        )
        return false
      }
    }

    return true
  }
}
