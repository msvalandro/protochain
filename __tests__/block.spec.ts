import { Block } from '../src/lib/block'

describe('Block tests', () => {
  it('should be valid', () => {
    const block = new Block(1, 'hash1', 'Block 1')

    expect(block.isValid()).toBe(true)
  })

  it('should not be valid due to index', () => {
    const block = new Block(-1, 'hash1', 'Block 1')

    expect(block.isValid()).toBe(false)
  })

  it('should not be valid due to previous hash', () => {
    const block = new Block(1, '', 'Block 1')

    expect(block.isValid()).toBe(false)
  })

  it('should not be valid due to data', () => {
    const block = new Block(1, '', '')

    expect(block.isValid()).toBe(false)
  })

  it('should create genesis block', () => {
    const block = Block.genesis()

    expect(block.isValid()).toBe(true)
    expect(block.getIndex()).toBe(0)
    expect(block.getHash()).toBe(
      'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
    )
  })
})
