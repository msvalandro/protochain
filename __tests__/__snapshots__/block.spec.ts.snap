// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block tests should be invalid if block not mined: Block not minaaaaed 1`] = `"Block not mined"`;

exports[`Block tests should be invalid if block not mined: Block not mined 1`] = `"Block not mined"`;
